src/runtime/virtcontainers/kata_agent.go

```diff
@@ -2241,20 +2241,19 @@ func (k *kataAgent) sendReq(spanCtx context.Context, request interface{}) (inter
 	msgName := string(proto.MessageName(request.(proto.Message)))
 
 	k.Lock()
+	client := k.client
+	reqHandlers := k.reqHandlers
+	k.Unlock()
 
-	if k.reqHandlers == nil {
-		k.Unlock()
+	if reqHandlers == nil || client == nil {
 		return nil, errors.New("Client has already disconnected")
 	}
 
-	handler := k.reqHandlers[msgName]
+	handler := reqHandlers[msgName]
```

Comment:
I think this operation should be done within the lock as well.
Assigning the map to `reqHandlers` with `reqHandlers := k.reqHandlers` is still a shallow copy, pointing to the original map. If you want to protect read/writes to the map, maybe you should retrieve the handler within the lock itself.

Reply:
Thx. There is no concurrent reading and writing in the map:reqHandlers, as we can see in func sendReq(). Here, I think it is necessary to protect k.reqHandlers param  itself  only. That is to say, setting k.reqHandlers value to be nil and getting k.reqHandlers param value are concurrent. However, reading/writing to the map is serial, as shown below:
![image](https://github.com/user-attachments/assets/499d5e4b-d55c-4b57-b489-e5b92401f8c6)